src/lib.rs

```diff
@@ -7,61 +7,278 @@
 //! ## Example
 //!
 //! ```rust
-//! use netlink_tc as tc;
+//! use netlink_packet_core::NetlinkMessage;
+//! use netlink_packet_route::RtnlMessage;
+//! use netlink_tc::ParseOptions;
 //!
-//! // Get list of qdiscs
-//! let qdiscs = tc::qdiscs::<tc::Netlink>().unwrap();
+//! // Retrieve netlink messages using `netlink-packet-route`.
+//! // See `examples` for more details.
+//! let messages: Vec<NetlinkMessage<RtnlMessage>> = vec![]; // init with netlink messages
 //!
-//! // Get list of classes
-//! let classes = tc::classes::<tc::Netlink>().unwrap();
+//! // Get list of tc qdiscs or classes
+//! let qdiscs = ParseOptions::new()
+//!     .fail_on_unknown_netlink_message(true)
+//!     .tc(messages.clone()).unwrap();
 //!
-//! // Get class for given interface
-//! let class = tc::class::<tc::Netlink>("eth0").unwrap();
+//! // Get list of links
+//! let links = ParseOptions::new()
+//!     .links(messages.clone()).unwrap();
 //! ```
+use netlink_packet_core::{NetlinkMessage, NetlinkPayload};
+use netlink_packet_route::{
+    link as netlink_link, tc as netlink_tc, LinkMessage as NlLinkMessage, RtnlMessage,
+    TcMessage as NlTcMessage,
+};
+use netlink_packet_utils::{nla::Nla, Emitable};
+
+use errors::TcError;
+use types::{Link, LinkAttr, LinkHeader, LinkMsg, Tc, TcAttr, TcHeader, TcMsg, TcOption, TcStats2};
 
 pub mod errors;
-pub mod link;
-pub mod tc;
 pub mod types;
 
 mod class;
 mod constants;
-mod netlink;
+mod link;
 mod qdiscs;
+mod tc;
 
 #[cfg(test)]
 mod test_data;
 #[cfg(test)]
 mod tests;
 
-pub use class::*;
-pub use netlink::*;
-pub use qdiscs::*;
-pub use types::*;
+/// Possible message types for `tc` messages.
+/// A subset of `rtnl::RtnlMessage` enum.
+pub enum RtNetlinkMessage {
+    GetQdisc(TcMsg),  /* RTM_GETQDISC */
+    GetClass(TcMsg),  /* RTM_GETCLASS */
+    GetLink(LinkMsg), /* RTM_GETLINK */
+}
 
-/// Get list of all `tc` qdiscs and classes.
-pub fn tc_stats<T: netlink::NetlinkConnection>() -> Result<Vec<Tc>, errors::TcError> {
-    tc::tc_stats::<T>()
+/// `OpenOptions` provides options for controlling how `netlink-tc` parses netlink messages.
+/// By default, unknown attributes and options are ignored.
+#[derive(Debug, Default)]
+pub struct ParseOptions {
+    fail_on_unknown_netlink_message: bool,
+    fail_on_unknown_attribute: bool,
+    fail_on_unknown_option: bool,
 }
 
-/// Get list of all `tc` qdiscs.
-pub fn qdiscs<T: netlink::NetlinkConnection>() -> Result<Vec<Tc>, errors::TcError> {
-    tc::qdiscs::<T>()
+impl ParseOptions {
+    /// Creates a new set of options with all flags set to false.
+    pub fn new() -> Self {
+        Self::default()
```

Comment:
Defaults in this case should be strict. Defaults are important and systems software should generally try to be conservative

Reply:
you mean we should set false for each field instead of using `Default`?

Reply:
Oh sorry, I meant we should set each field to True

Reply:
while that makes sense, setting true right now would mean the call will definitely fail unless we add parsing for the ignored fields. some fields in the response like `HwOffload`, or qdiscs other than `fq_codel` and `htb` are currently not parsed and setting to true means the call will keep failing until the library is "complete".

does it still make sense to set them to true? maybe we can do that eventually?

Reply:
I think it still makes sense. The user cannot easily tell what is implemented and what is not.

Reply:
okay, I'll add a comment calling that out in the doc